src/proxy_lib/proxy_lib.c

```diff
@@ -0,0 +1,253 @@
+ * by umfMemoryProviderCreate() and umfPoolCreate().
+ * That memory will be leaked on purpose (we believe OS will free it),
+ * because we cannot free the memory containing array of destructors
+ * that have to be called at the end.
+ * It uses 892 kB on Linux.
+ *
+ * Ad 2)
+ * The UMF pool allocator (the main one) is used from the creation to the destruction
+ * of a UMF pool to allocate memory needed by an application. It should be freed
+ * by an application.
+ */
+
+static UTIL_ONCE_FLAG Base_alloc_leak_initialized = UTIL_ONCE_FLAG_INIT;
+static umf_ba_linear_pool_t *Base_alloc_leak = NULL;
+static umf_memory_provider_handle_t OS_memory_provider = NULL;
+static umf_memory_pool_handle_t Proxy_pool = NULL;
+
+// it protects us from recursion in umfPool*()
+static __TLS int was_called_from_umfPool = 0;
+
+/*****************************************************************************/
+/*** The constructor and destructor of the proxy library *********************/
+/*****************************************************************************/
+
+void proxy_lib_create_common(void) {
+    umf_os_memory_provider_params_t os_params =
+        umfOsMemoryProviderParamsDefault();
+    enum umf_result_t umf_result;
+
+    umf_result = umfMemoryProviderCreate(umfOsMemoryProviderOps(), &os_params,
+                                         &OS_memory_provider);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating OS memory provider failed\n");
+        exit(-1);
+    }
+
+    umf_result = umfPoolCreate(umfJemallocPoolOps(), OS_memory_provider, NULL,
+                               0, &Proxy_pool);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating jemalloc pool manager failed\n");
+        exit(-1);
+    }
+    // The UMF pool has just been created (Proxy_pool != NULL). Stop using
+    // the linear allocator and start using the UMF pool allocator from now on.
+}
+
+void proxy_lib_destroy_common(void) {
+    umf_memory_pool_handle_t pool = Proxy_pool;
+    Proxy_pool = NULL;
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(OS_memory_provider);
+    // the "LEAK" linear base allocator is not destroyed at all
+}
+
+/*****************************************************************************/
+/*** Generic version of realloc() of linear base allocator *******************/
+/*****************************************************************************/
+
+static inline void *ba_generic_realloc(umf_ba_linear_pool_t *pool, void *ptr,
+                                       size_t size) {
+    if (size == 0) {
+        // it means free(ptr), but linear base allocator does not implement free()
+        return NULL;
+    }
+
+    assert(ptr); // it is verified in the main realloc()
+
+    size_t max_size = umf_ba_linear_pool_contains_pointer(pool, ptr);
+    assert(max_size > 0); // assert that pool contains the pointer
+
+    void *new_ptr = umf_ba_linear_alloc(pool, size);
```

Comment:
There's a possible race here if some other thread calls `umf_ba_linear_pool_contains_pointer` after this thread calls `umf_ba_linear_pool_contains_pointer` and before `umf_ba_linear_alloc`.

Reply:
1) The linear allocator is used mainly in the phase of loading the application in the single-threaded environment.
2) It can be used in the MT environment only in the case of "recursive" calls when memory is allocated inside of the external libraries (hwloc for example). In such case I do not see anything dangerous in this race, because the linear allocator has an internal lock. If you see a dangerous scenario, explain it in details, please.

Reply:
Ok, you're right - I misunderstood how this code is used.